From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Wed, 2 Jul 2025 22:39:21 +0800
Subject: [PATCH] Fix error throwing in configuration state handlers
 disconnecting logics

A possible fix for https://github.com/PaperMC/Folia/issues/378
Should be a new way to gracefully and more correctly retire the entity scheduler of players

diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 31101f5af05772a0842b56ecc2dd8c79eebee933..3332b32d93f3def3ed266022660043f431777b09 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -98,6 +98,20 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         }
         // Folia start - region threading
         if (!(this instanceof ServerGamePacketListenerImpl gamePacketListener)) {
+            // Luminol start - Fix error throwing in configuration state handlers disconnecting logics
+            // we are not starting to turn into a game packet handler, so we can safely return
+            if (this instanceof ServerConfigurationPacketListenerImpl && !io.papermc.paper.threadedregions.RegionizedServer.isNotOwnedByGlobalRegion(this.connection)) {
+                // folia's weird entity scheduler process?
+                if (this.connection.savedPlayerForLoginEventLegacy != null) {
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueTickTaskQueue(
+                            this.connection.savedPlayerForLoginEventLegacy.level(),
+                            this.connection.savedPlayerForLoginEventLegacy.getBlockX() >> 4,
+                            this.connection.savedPlayerForLoginEventLegacy.getBlockZ() >> 4,
+                            () -> this.connection.savedPlayerForLoginEventLegacy.getBukkitEntity().taskScheduler.retire());
+                }
+                return;
+            }
+            // Luminol end
             throw new UnsupportedOperationException("Configuration mode is not supported in its current state");
         }
         gamePacketListener.player.getBukkitEntity().taskScheduler.retire();
@@ -356,14 +370,25 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
             return;
         }
         // Folia start - region threading
-        if (!(this instanceof ServerGamePacketListenerImpl gamePacketListener)) {
-            throw new UnsupportedOperationException("Configuration mode is not supported in its current state");
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            // Luminol start - Fix error throwing in configuration state handlers disconnecting logics
+            // folia's weird entity scheduler process?
+            // we'll retire soonly
+            //throw new UnsupportedOperationException("Configuration mode is not supported in its current state");
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(gamePacketListener.player)) { // Folia - region threading
+                this.connection.disconnectSafely(disconnectionDetails); // Folia - region threading - it HAS to be delayed/async to avoid deadlock if we try to wait for another region
+                return;
+            }
+        }else {
+            // disconnecting configuration handler off globalregion
+            if (!io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread()) {
+                // calling on configuration mode
+                this.connection.disconnectSafely(disconnectionDetails); // Folia - region threading - it HAS to be delayed/async to avoid deadlock if we try to wait for another region
+                return;
+            }
         }
         // Folia end - region threading
-        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(gamePacketListener.player)) { // Folia - region threading
-            this.connection.disconnectSafely(disconnectionDetails); // Folia - region threading - it HAS to be delayed/async to avoid deadlock if we try to wait for another region
-            return;
-        }
+        // Luminol end
 
         Component reason;
         Component leaveMessage;
